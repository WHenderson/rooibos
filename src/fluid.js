const State = require('./state');

const state = new State();

function combine(name, func, ...options) {
    const final = Object.assign({}, ...options, { name, func });

    if (!final.func) {
        final.func = final.name;
        if (!final.func)
            throw new Error('Missing required');

        final.name = final.func.name || '';
    }

    return final;
}

const testishAlias = testish;

function testish(defaults) {

    function testish(options) {
        return testishAlias(Object.assign({}, defaults, options));
    }

    function describe(name, func, options) {
        return state.describe(combine(name, func, defaults, options));
    }

    describe.only = function only(name, func, options) {
        return state.test(combine(name, func, defaults, options, { only: true }));
    };

    describe.skip = function only(name, func, options) {
        return state.test(combine(name, func, defaults, options, { skip: true }));
    };

    function test(name, func, options) {
        return state.test(combine(name, func, defaults, options));
    }

    test.only = function only(name, func, options) {
        return state.test(combine(name, func, defaults, options, { only: true }));
    };

    test.skip = function only(name, func, options) {
        return state.test(combine(name, func, defaults, options, { skip: true }));
    };

    function before(name, func, options) {
        return state.before(combine(name, func, defaults, options));
    }

    function after(name, func, options) {
        return state.before(combine(name, func, defaults, options));
    }

    function beforeEach(name, func, options) {
        return state.before(combine(name, func, defaults, options));
    }

    function afterEach(name, func, options) {
        return state.before(combine(name, func, defaults, options));
    }

    return {
        testish,
        before,
        beforeEach,
        describe,
        test,
        afterEach,
        after
    }
}

module.exports = testish({});

// timeout(ms).describe(name, func)
// timeout(ms).describe.only(name, func)
// timeout(ms).describe.skip(name, func)
// timeout(ms).test(name, func)
// timeout(ms).test.only(name, func)
// timeout(ms).test.skip(name, func)
// should timeouts be only for tests, or for everything including describe and hooks?
// just return a new object from each fluid function, dont use classes