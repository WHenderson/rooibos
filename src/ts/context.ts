export interface Cancel {
    (this: Context, context: Context) : Promise<void> | void;
}

export interface Context {
    name: string;
    parents: Context[];

    cancel?: Cancel;
}
