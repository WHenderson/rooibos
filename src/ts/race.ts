export function race(timeout: number, pending : Promise<void>) : Promise<void> {
    if (!timeout)
        return pending;

    let id;

    return Promise
        .race([
            pending,
            new Promise((resolve, reject) => {
                id = setTimeout(() => {
                    clearTimeout(id);
                    reject(new Error('Timeout'));
                })
            })
        ])
        .then(() => {
            clearTimeout(id);
        });
}