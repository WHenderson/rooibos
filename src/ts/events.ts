import { Context } from "./context";

export enum EventType {
    ENTER = 'ENTER',
    SKIP = 'SKIP',
    PENDING = 'PENDING',
    SUCCESS = 'SUCCESS',
    FAILURE = 'FAILURE',
    TIMEOUT = 'TIMEOUT',
    ABORT = 'ABORT',
    LEAVE = 'LEAVE'
}

export enum NodeType {
    describe = 'describe',
    test = 'test'
}

export enum HookType {
    before = 'before',
    beforeEach = 'beforeEach',
    afterEach = 'afterEach',
    after = 'after'
}

export type EntryType = NodeType | HookType;
export const EntryType = { ...NodeType, ...HookType };

export interface Event {
    name : string,
    entry : EntryType,
    event : EventType,
    context : Context,
    exception? : Error
}