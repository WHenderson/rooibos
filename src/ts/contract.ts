import { Context } from "./context";
import {EntryType} from "./events";

export interface Action extends Function {
    (this: Context, context?: Context): Promise<void> | void;
}

export interface Options {
    only?: boolean;
    skip?: boolean;

    timeoutDefault?: number;
    timeoutBefore?: number;
    timeoutBeforeEach?: number;
    timeoutAfterEach?: number;
    timeoutAfter?: number;
    timeoutDescribe?: number;
    timeoutTest?: number;
}

export const InheritableOptions = [
    'timeoutDefault',
    'timeoutBefore',
    'timeoutBeforeEach',
    'timeoutAfterEach',
    'timeoutAfter',
    'timeoutDescribe',
    'timeoutTest'
];

export function resolveTimeout(options: Options, entry: EntryType) : number {
    const specific = (() => {
        switch (entry) {
            case EntryType.before:
                return options.timeoutBefore;
            case EntryType.beforeEach:
                return options.timeoutBeforeEach;
            case EntryType.afterEach:
                return options.timeoutAfterEach;
            case EntryType.after:
                return options.timeoutAfter;
            case EntryType.describe:
                return options.timeoutDescribe;
            case EntryType.test:
                return options.timeoutTest;
        }
    })();

    return (typeof specific !== 'number' ? options.timeoutDefault : specific) || 0;
}

export interface NamedOptions extends Options {
    name: string;
    callback: Action;
}

interface NodeContract {
    (name: string, callback : Action, options?: Options) : Promise<void>;
    (callback : Action, options?: Options) : Promise<void>;
}

interface HookContract {
    (name: string, callback : Action, options?: Options) : void;
    (callback : Action, options?: Options) : void;
}

export interface Describe extends NodeContract {
    only: NodeContract;
    skip: NodeContract;
}

export interface Test extends NodeContract {
    only: NodeContract;
    skip: NodeContract;
}

export interface Before extends HookContract {
}

export interface BeforeEach extends HookContract {
}

export interface After extends HookContract {
}

export interface AfterEach extends HookContract {
}