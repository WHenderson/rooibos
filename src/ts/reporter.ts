import {Event} from "./events";

export interface Reporter {
    on(event : Event) : Promise<void>;
}

