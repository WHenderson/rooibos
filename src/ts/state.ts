import {Context} from './context';
import {InheritableOptions, NamedOptions, Options, resolveTimeout} from "./contract";
import {EntryType, EventType, HookType, NodeType} from "./events";
import {strict as assert} from 'assert';
import {Reporter} from "./reporter";
import {VerboseReporter} from "./reporters/verbose";
import {race} from "./race";

// TODO: When running actions, they need to be in a race where one contestant is an unresolved which can be used to abort (when cancelling)

enum Status {
    Ready,
    Started,
    Cancelled
}

function isStatusMutable(status: Status) : boolean {
    return status === Status.Ready;
}

interface StateResolve {
    () : void
}
interface StateReject {
    () : void
}

interface StateCancel {
    (this: State, state?:State) : Promise<void> | void;
}
interface StateAbort {
    (this: State, state?:State) : Promise<void> | void;
}

interface State {
    status: Status;

    filteredChildren: boolean;
    finalChild: NamedOptions; // used to identify the last child in the queue
    entry: NodeType;
    options: Options;
    context: Context;
    actions: Promise<void>;

    before: NamedOptions[];
    beforeEach: NamedOptions[];
    afterEach: NamedOptions[];
    after: NamedOptions[];

    // Cancel just these actions etc
    cancel: StateCancel;
    abort?: StateAbort;
}

export class StateStack {
    private readonly _root : State;
    private readonly _stack : State[];
    private readonly _reporter : Reporter;

    constructor(reporter?: Reporter) {
        this._stack = [];
        this.push(NodeType.describe, { name: 'root', callback: undefined });
        this._reporter = reporter || new VerboseReporter();
    }

    private push(entry: NodeType, options: NamedOptions) : State {
        assert(this._stack.length === 0 || this.current.entry === EntryType.describe, 'Can only nest inside describe blocks');

        const state : State = {
            status: Status.Ready,
            entry: entry,
            filteredChildren: false,
            finalChild: undefined,
            options: options,
            context: {
                name: options.name,
                parents: this._stack.map(state => state.context)
            },
            actions: Promise.resolve(),

            before: [],
            beforeEach: [],
            afterEach: [],
            after: [],

            cancel: async () => {
                if (state.status === Status.Cancelled)
                    return;

                state.status = Status.Cancelled;

                if (state.context.cancel) {
                    const cancel = state.context.cancel;
                    delete state.context.cancel;

                    await cancel.call(state.context, state.context);
                }

                if (state.abort) {
                    const abort = state.abort;
                    delete state.abort;

                    await abort.call(state, state);
                }

                //await state.actions; // Actions in the ready state will just cancel without running sub tasks or hooks etc
            }
        };

        this._stack.push(state);

        return state;
    }

    private pop(state: State) : State {
        assert(this.current === state, `Popping incorrect state (Found '${this.current.context.name}', expected '${state.context.name}')`);
        return this._stack.pop();
    }

    private get current() : State {
        return this._stack[this._stack.length - 1];
    }

    private get parent() : State {
        if (this._stack.length >= 2)
            return this._stack[this._stack.length - 2];
        else
            return null;
    }

    private hook(entry: HookType, options: NamedOptions) : void {
        if (this.current.status !== Status.Ready)
            throw new Error('Hooks have already started, cannot add more at this point');

        this.current[entry].push(options);
    }

    public before(options: NamedOptions) : void {
        this.hook(HookType.before, options);
    }
    public beforeEach(options: NamedOptions) : void {
        this.hook(HookType.beforeEach, options);
    }
    public afterEach(options: NamedOptions) : void {
        this.hook(HookType.afterEach, options);
    }
    public after(options: NamedOptions) : void {
        this.hook(HookType.after, options);
    }

    private async _cancel(state: State) {
        for (let s of this._stack.slice().reverse()) {
            if (s.status !== Status.Cancelled)
                await s.cancel(s);

            if (s === state)
                break;
        }
    }

    private async _runHooks(state: State, entry: HookType, context?: Context) : Promise<void> {
        const actions = state[entry];
        context = context || state.context;

        for (let options of actions) {
            await this._reporter.on({ event: EventType.ENTER, entry: entry, context, name: options.name });
            try {
                if (state.status !== Status.Cancelled) {
                    const waitAbort: Promise<void> = new Promise((resolve, reject) => {
                        state.abort = () => reject(new Error('Abort'));
                    });
                    let timeoutId;
                    const timeout = resolveTimeout(options, entry);
                    const waitTimeout: Promise<void> = timeout
                        ? new Promise((resolve, reject) => {
                            timeoutId = setTimeout(() => {
                                reject(new Error('Timeout'));
                            })
                        })
                        : undefined;

                    try {
                        //await race(options.callback.call(context, context), resolveTimeout(options, entry), waitAbort);

                        const waitAction = options.callback.call(context, context);

                        if (timeout)
                            await Promise.race([waitAbort, waitTimeout, waitAction]);
                        else
                            await Promise.race([waitAbort, waitAction]);
                    } finally {
                        clearTimeout(timeoutId);
                        delete state.abort;
                    }

                    await this._reporter.on({event: EventType.SUCCESS, entry: entry, context, name: options.name});
                }
                else {
                    await this._reporter.on({event: EventType.SKIP, entry: entry, context, name: options.name});
                }
            }
            catch (ex) {
                if (ex && ex.message === 'Abort') {
                    await this._reporter.on({ event: EventType.ABORT, entry: entry, context, name: options.name })
                }
                else if (ex && ex.message === 'Timeout') {
                    await this._cancel(state);
                    await this._reporter.on({ event: EventType.TIMEOUT, entry: entry, context, name: options.name })
                }
                else {
                    await this._reporter.on({ event: EventType.FAILURE, entry: entry, context, name: options.name })
                }
            }
            finally {
                await this._reporter.on({ event: EventType.LEAVE, entry: entry, context, name: options.name })
            }
        }
    }
    private async _queueNode(options: NamedOptions, entry: NodeType) : Promise<void> {
        // compile options
        options = Object.assign(
            {},
            // only include inheritable options
            Object
                .keys(this.current.options)
                .filter(key => InheritableOptions.includes(key))
                .reduce((obj, key) => {
                    obj[key] = this.current.options[key];
                    return obj;
                }, {}),
            options
        );

        if (this.current.entry !== NodeType.describe)
            throw new Error('Cannot nest here');

        if (options.only) {
            if (!isStatusMutable(this.current.status) && !this.current.filteredChildren)
                throw new Error('Actions have already started, can not use .only at this point');

            this.current.filteredChildren = true;
        }

        // tag this node as the final child
        this.current.finalChild = options;

        return this.current.actions = this.current.actions.then(async () => {
            const parent = this.current;

            const skip = parent.status == Status.Cancelled || options.skip || (parent.filteredChildren && !options.only);

            // first non-skipped child?
            if (!skip && parent.status == Status.Ready) {
                parent.status = Status.Started;
                await this._runHooks(parent, HookType.before);
            }

            const me = this.push(entry, options);
            try {
                if (!skip) {
                    await this._runHooks(parent, HookType.beforeEach, me.context);
                }

                await this._reporter.on({ event: EventType.ENTER, entry: entry, context: me.context, name: options.name });

                if (!skip) {

                    let waitAbort = new Promise((resolve, reject) => {
                        me.abort = async () => { reject(new Error('Abort')); }
                    });
                    let timeoutId;
                    const timeout = resolveTimeout(options, entry);
                    const waitTimeout: Promise<void> = timeout
                        ? new Promise((resolve, reject) => {
                            timeoutId = setTimeout(() => {
                                reject(new Error('Timeout'));
                            })
                        })
                        : undefined;

                    try {
                        // Run the immediate action
                        try {
                            const waitAction = options.callback.call(me.context, me.context);

                            if (timeout)
                                await Promise.race([waitAbort, waitTimeout, waitAction]);
                            else
                                await Promise.race([waitAbort, waitAction]);
                        }
                        catch (ex) {
                            if (ex && ex.message === 'Abort') {
                                await this._reporter.on({ event: EventType.ABORT, entry: entry, context: me.context, name: options.name });
                            }
                            else if (ex && ex.message === 'Timeout') {
                                await me.cancel(me);
                                await this._reporter.on({ event: EventType.TIMEOUT, entry: entry, context: me.context, name: options.name })
                            }
                            else {
                                await this._reporter.on({
                                    event: EventType.FAILURE,
                                    entry: entry,
                                    context: me.context,
                                    exception: ex,
                                    name: options.name
                                });
                            }
                        }
                        finally {
                            delete me.abort;
                        }

                        // Notify we are running the children
                        // Don't notify of children since some may have been waited on and thus executed earlier
                        //await this._reporter.on({event: EventType.PENDING, entry: entry, context: me.context});

                        // Run children
                        try {
                            if (timeout && me.status !== Status.Cancelled)
                                await Promise.race([waitTimeout, me.actions]);
                            else
                                await me.actions;
                        }
                        catch (ex) {
                            if (ex && ex.message === 'Timeout') {
                                for (let s of this._stack.slice().reverse()) {
                                    if (s.status !== Status.Cancelled)
                                        await s.cancel(s);

                                    await s.actions;

                                    if (s === me)
                                        break;
                                }
                                await this._reporter.on({ event: EventType.TIMEOUT, entry: entry, context: me.context, name: options.name })
                            }
                        }

                        if (me.status !== Status.Cancelled)
                            await this._reporter.on({ event: EventType.SUCCESS, entry: entry, context: me.context, name: options.name });
                    }
                    finally {
                        clearTimeout(timeoutId);
                    }
                }
                else {
                    await this._reporter.on({ event: EventType.SKIP, entry: entry, context: me.context, name: options.name });
                }

                await this._reporter.on({ event: EventType.LEAVE, entry: entry, context: me.context, name: options.name });

                if (!skip) {
                    await this._runHooks(parent, HookType.afterEach, me.context);
                }
            }
            finally {
                this.pop(me);
            }

            // last child
            if (parent.status !== Status.Ready && parent.finalChild === options) {
                await this._runHooks(parent, HookType.after);
            }
        });
    }

    public describe(options: NamedOptions) : Promise<void> {
        return this._queueNode(options, NodeType.describe);
    }
    public test(options: NamedOptions) : Promise<void> {
        return this._queueNode(options, NodeType.test);
    }
}