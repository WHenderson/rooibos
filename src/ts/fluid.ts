import { Action, Options, Describe, After, AfterEach, Before, BeforeEach, NamedOptions, Test } from "./contract";
import {StateStack} from "./state";

interface Testish {
    testish: (defaults: Options) => Testish;
    describe: Describe;
    test: Test;
    before: Before;
    beforeEach: BeforeEach;
    afterEach: AfterEach;
    after: After;
}

function isString(x: any): x is string {
    return typeof x === "string";
}

function combine(name : string | Action, callback: Action | Options, ...options : Options[]) : NamedOptions {
    if (isString(name)) {
        return Object.assign(
            {},
            ...options,
            { name, callback}
        );
    }
    else {
        return Object.assign(
            {},
            callback,
            ...options,
            {
                name: name.name || '',
                callback: name
            }
        )
    }
}

const stateStack = new StateStack();

function testish(defaults? : Options) : Testish {
    defaults = defaults || {};

    function after(name: string, callback : Action, options?: Options) : void;
    function after(callback : Action, options?: Options) : void;
    function after(name: string | Action, callback : Action | Options, options?: Options) : void {
        stateStack.after(combine(name, callback, defaults, options));
    }

    function afterEach(name: string, callback : Action, options?: Options) : void;
    function afterEach(callback : Action, options?: Options) : void;
    function afterEach(name: string | Action, callback : Action | Options, options?: Options) : void {
        stateStack.afterEach(combine(name, callback, defaults, options));
    }

    function before(name: string, callback : Action, options?: Options) : void;
    function before(callback : Action, options?: Options) : void;
    function before(name: string | Action, callback : Action | Options, options?: Options) : void {
        stateStack.before(combine(name, callback, defaults, options));
    }

    function beforeEach(name: string, callback : Action, options?: Options) : void;
    function beforeEach(callback : Action, options?: Options) : void;
    function beforeEach(name: string | Action, callback : Action | Options, options?: Options) : void {
        stateStack.beforeEach(combine(name, callback, defaults, options));
    }

    function describe(name: string, callback : Action, options?: Options) : Promise<void>;
    function describe(callback : Action, options?: Options) : Promise<void>;
    function describe(name: string | Action, callback : Action | Options, options?: Options) : Promise<void> {
        return stateStack.describe(combine(name, callback, defaults, options));
    }

    function describeOnly(name: string, callback : Action, options?: Options) : Promise<void>;
    function describeOnly(callback : Action, options?: Options) : Promise<void>;
    function describeOnly(name: string | Action, callback : Action | Options, options?: Options) : Promise<void> {
        return stateStack.describe(combine(name, callback, defaults, options, { only: true }));
    }

    function describeSkip(name: string, callback : Action, options?: Options) : Promise<void>;
    function describeSkip(callback : Action, options?: Options) : Promise<void>;
    function describeSkip(name: string | Action, callback : Action | Options, options?: Options) : Promise<void> {
        return stateStack.describe(combine(name, callback, defaults, options, { skip: true }));
    }
    
    describe.only = describeOnly;
    describe.skip = describeSkip;

    function test(name: string, callback : Action, options?: Options) : Promise<void>;
    function test(callback : Action, options?: Options) : Promise<void>;
    function test(name: string | Action, callback : Action | Options, options?: Options) : Promise<void> {
        return stateStack.test(combine(name, callback, defaults, options));
    }

    function testOnly(name: string, callback : Action, options?: Options) : Promise<void>;
    function testOnly(callback : Action, options?: Options) : Promise<void>;
    function testOnly(name: string | Action, callback : Action | Options, options?: Options) : Promise<void> {
        return stateStack.test(combine(name, callback, defaults, options, { only: true }));
    }

    function testSkip(name: string, callback : Action, options?: Options) : Promise<void>;
    function testSkip(callback : Action, options?: Options) : Promise<void>;
    function testSkip(name: string | Action, callback : Action | Options, options?: Options) : Promise<void> {
        return stateStack.test(combine(name, callback, defaults, options, { skip: true }));
    }

    test.only = testOnly;
    test.skip = testSkip;    
    
    return {
        after,
        afterEach,
        before,
        beforeEach,
        describe,
        test,
        testish: function (options: Options) {
            return testish(Object.assign({}, defaults, options));
        }
    }
}

const root = testish();

export const
    before = root.before,
    beforeEach = root.beforeEach,
    describe = root.describe,
    test = root.test,
    afterEach = root.afterEach,
    after = root.after;
