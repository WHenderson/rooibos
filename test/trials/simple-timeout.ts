import { before, beforeEach, describe, test, afterEach, after, testish } from '../../src/ts/fluid';
import { strict as assert } from 'assert';

before('my-before', async () => {
});

beforeEach('my-beforeEach', async () => {
});

afterEach('my-afterEach', async () => {
});

after('my-after', async () => {
});

testish({ timeoutDescribe: 500 }).describe('outer', async () => {
    let x = 0;

    before('my-inner-before', async () => {
    });

    beforeEach('my-inner-beforeEach', async () => {
    });

    afterEach('my-inner-afterEach', async () => {
    });

    after('my-inner-after', async () => {
    });

    test('a', async () => {
        console.log('1');
    });

    test('b', async (context) => {

        await new Promise((resolve, reject) => {
            const id = setTimeout(resolve, 500);
            context.cancel = () => {
                clearTimeout(id);
                //reject(new Error('hmm'));
            }
        });

        console.log('2');
    });

    test('c', async (context) => {

        await new Promise((resolve, reject) => {
            const id = setTimeout(resolve, 500);
            context.cancel = () => {
                clearTimeout(id);
            }
        });

        console.log('3');
    });

    describe('nest', () => {
        test('x', () => {

        });
        test('y', () => {

        });
    })

});