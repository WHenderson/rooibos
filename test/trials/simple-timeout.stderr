(node:2004) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'message' of undefined
    at current.actions.current.actions.then (C:\Code\OwnGithub\testish\src\ts\state.js:160:32)
(node:2004) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:2004) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:2004) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'message' of undefined
    at current.actions.current.actions.then (C:\Code\OwnGithub\testish\src\ts\state.js:160:32)
(node:2004) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:2004) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'message' of undefined
    at current.actions.current.actions.then (C:\Code\OwnGithub\testish\src\ts\state.js:160:32)
(node:2004) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:2004) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'message' of undefined
    at current.actions.current.actions.then (C:\Code\OwnGithub\testish\src\ts\state.js:160:32)
(node:2004) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
